web/src/reducers/budget.js

```diff
@@ -250,113 +254,172 @@ const buildBudget = bigState => {
-          newState[fs].combined.total.state += cost.stateShare;
-        };
+        // Plus the subtotals for the cost categories (i.e., the
+        // Medicaid share)
+        newState[fs].combined[year].federal += costShares.fedShare;
+        newState[fs].combined.total.federal += costShares.fedShare;
 
-        addShares(fundingSource);
-
-        if (fundingSource === 'hie' || fundingSource === 'hit') {
-          // We need to track expenses as combined between HIE and HIT,
-          // so for those funding sources, run again.
-          addShares('hitAndHie');
-        } else {
-          // But for MMIS activities, we need to track costs per FFP level
-          // (90/10, 75/25, 50/50)
-          const ffpLevel = `${allocation[year].ffp.federal}-${
-            allocation[year].ffp.state
-          }`;
-
-          addShares('mmisByFFP', ffpLevel);
-
-          // MMIS-by-FFP has additional totals that the others don't have
-          newState.mmisByFFP[ffpLevel][year].total += totalCost;
-          newState.mmisByFFP[ffpLevel].total.total += totalCost;
-          newState.mmisByFFP.combined[year].total += totalCost;
-          newState.mmisByFFP.combined.total.total += totalCost;
-        }
+        newState[fs].combined[year].state += costShares.stateShare;
+        newState[fs].combined.total.state += costShares.stateShare;
+      };
+
+      updateCosts(fundingSource);
+
+      newState.combined[year].federal += costShares.fedShare;
+      newState.combined.total.federal += costShares.fedShare;
+      newState.combined[year].state += costShares.stateShare;
+      newState.combined.total.state += costShares.stateShare;
+
+      if (fundingSource === 'hie' || fundingSource === 'hit') {
+        // We need to track HIE and HIT combined, so we have a funding source
+        // called hitAndHie.  If this is HIE or HIT, roll it into that one too.
+        updateCosts('hitAndHie');
+      } else {
+        // For MMIS, we need to track costs by "FFP type", too - that is,
+        // the cost allocation.  So we need to know total costs for MMIS
+        // at the 90/10 level, at the 75/25 level, and at the 50/50 level.
+
+        // Compute a string that represents the MMIS level for this activity
+        const ffpLevel = `${allocation[year].ffp.federal}-${
+          allocation[year].ffp.state
+        }`;
+
+        // Then do basically the same as updateCosts() above, but we only
+        // need to track subtotals, not individual cost categories
+        [ffpLevel, 'combined'].forEach(prop => {
+          newState.mmisByFFP[prop][year].federal += costShares.fedShare;
+          newState.mmisByFFP[prop].total.federal += costShares.fedShare;
+
+          newState.mmisByFFP[prop][year].state += costShares.stateShare;
+          newState.mmisByFFP[prop].total.state += costShares.stateShare;
+
+          newState.mmisByFFP[prop][year].total += totalCost;
+          newState.mmisByFFP[prop].total.total += totalCost;
```

Comment:
I was going to make an argument for making this less redundant, but it being so explicit makes it way easier to read (for me, anyway)!

Reply:
Back in the Before Times™ (first iteration), there were lots of loops and abstractions, but then when we needed to change how something was calculated it took days and days to figure out where that happened.  Literally, days.